In [1]:
import afscgap

# Basic Query
This first example queries for Pacific cod (*Pasiphaea pacifica*) in the Gulf of Alaska in 2021. The library will automatically generate HTTP queries, converting from Python types to [ORDS](https://www.oracle.com/database/technologies/appdev/rest.html) query syntax. This first example simply records the number of records returned.

In [2]:
import afscgap

query = afscgap.Query()
query.filter_year(eq=2021)
query.filter_srvy(eq='GOA')
query.filter_scientific_name(eq='Pasiphaea pacifica')

results = query.execute()

In [3]:
len(list(results))

52

# Iterating with a for loop
The easiest way to interact with results is a simple for loop. This next example determines the frequency of different catch per unit effort where Pacific cod were reported:


In [4]:
import afscgap

# Mapping from CPUE to count
count_by_cpue = {}

# Build query
query = afscgap.Query()
query.filter_year(eq=2021)
query.filter_srvy(eq='GOA')
query.filter_scientific_name(eq='Pasiphaea pacifica')
results = query.execute()

# Iterate through results and count
for record in results:
  cpue = record.get_cpue_weight(units='kg/ha')
  cpue_rounded = round(cpue)
  count = count_by_cpue.get(cpue_rounded, 0) + 1
  count_by_cpue[cpue_rounded] = count

# Print the result
print(count_by_cpue)

{0: 44, 1: 7, 2: 1}


### Iterating with functional programming
A for loop is not the only option for iterating through results. List comprehensions and other functional programming methods can be used as well.


In [5]:
import statistics

import afscgap

# Build query
query = afscgap.Query()
query.filter_year(eq=2021)
query.filter_srvy(eq='GOA')
query.filter_scientific_name(eq='Pasiphaea pacifica')
results = query.execute()

# Get temperatures in Celsius
temperatures = [record.get_bottom_temperature(units='c') for record in results]

# Take the median
print(statistics.median(temperatures))

5.3


# Load in Pandas
The results from the `afscgap` package are serializable and can be loaded into other tools like [Pandas](https://pandas.pydata.org/). This example loads Pacific cod from 2021 Gulf of Alaska into a data frame.


In [6]:
import pandas

import afscgap

query = afscgap.Query()
query.filter_year(eq=2021)
query.filter_srvy(eq='GOA')
query.filter_scientific_name(eq='Pasiphaea pacifica')
results = query.execute()

pandas.DataFrame(results.to_dicts())

year srvy                              survey  survey_id    cruise  \
0   2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
1   2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
2   2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
3   2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
4   2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
5   2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
6   2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
7   2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
8   2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
9   2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
10  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
11  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
12  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
13  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
14  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
15  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
16  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
17  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
18  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
19  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
20  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
21  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
22  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
23  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
24  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
25  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
26  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
27  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
28  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
29  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
30  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
31  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
32  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
33  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
34  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
35  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
36  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
37  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
38  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
39  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
40  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
41  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
42  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
43  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
44  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
45  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
46  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
47  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
48  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
49  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
50  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey       47.0  202101.0   
51  2021.0  GOA  Gulf of Alaska Bottom Trawl Survey   

# Advanced filtering
Queries so far have focused on filters requiring equality but range queries can be built as well.

In [7]:
import afscgap

# Build query
query = afscgap.Query()
query.filter_year(min_val=2015, max_val=2019)   # Note min/max_val
query.filter_srvy(eq='GOA')
query.filter_scientific_name(eq='Pasiphaea pacifica')
results = query.execute()

# Sum weight
weights = map(lambda x: x.get_weight(units='kg'), results)
total_weight = sum(weights)
print(total_weight)

58.94600000000002


### Zero-catch inference
Until this point, these examples use presence-only data. However, the `afscgap` package can infer negative or "zero catch" records as well.


In [8]:
import afscgap

# Mapping from CPUE to count
count_by_cpue = {}

# Build query
query = afscgap.Query()
query.filter_year(eq=2021)
query.filter_srvy(eq='GOA')
query.filter_scientific_name(eq='Pasiphaea pacifica')
query.set_presence_only(False)  # Added to earlier example
results = query.execute()

# Iterate through results and count
for record in results:
  cpue = record.get_cpue_weight(units='kg/ha')
  cpue_rounded = round(cpue)
  count = count_by_cpue.get(cpue_rounded, 0) + 1
  count_by_cpue[cpue_rounded] = count

# Print the result
print(count_by_cpue)

/home/sam/.local/lib/python3.10/site-packages/afscgap/__init__.py:1222: UserWarning: Your query may return a very large amount of records. Be sure to interact with results in a memory efficient way.
  warn_function = lambda x: warnings.warn(x)


{0: 521, 1: 7, 2: 1}


This example revisits the earlier snippet for CPUE counts but `set_presence_only(False)` directs the library to look at additional data on hauls, determining which hauls did not have Pacific cod. This lets the library return records for hauls in which Pacific cod were not found. This can be seen in differences in counts reported:

| Rounded CPUE | Count with set_presence_only(True) | Count with set_presence_only(False) |
|--------------|------------------------------------|-------------------------------------|
| 0 kg/ha      | 44                                 | 521                                 |
| 1 kg/ha      | 7                                  | 7                                   |
| 2 kg/ha      | 1                                  | 1                                   |

Put simply, while the earlier example showed CPUE counts for hauls in which Pacific cod were seen, this revised example reports for all hauls in the Gulf of Alaska in 2021.


# More information
Please see the [API documentation](https://pyafscgap.org/devdocs/afscgap.html) for the Python library for additional details.